File: libvitriol/hermetis/retrieval.py
"""
VITRIOL Emulated Memory — Retrieval with Cascading (Spreading Activation)

Multi-hop retrieval: direct search → edge traversal → score → rank.
When VITRIOL_SEMANTIC_MODE=on, relevance scoring uses cosine similarity
via sentence-transformers instead of keyword overlap.
"""

import math
import os
import re
from typing import Optional

from . import compact, db
from .scorer import keyword_overlap, recency_score, compute_score, estimate_tokens


# Default scoring weights — overridable via env or config
DEFAULT_TOP_K = int(os.environ.get('MEMORY_TOP_K', '5'))
DEFAULT_CASCADE_DEPTH = int(os.environ.get('MEMORY_CASCADE_DEPTH', '1'))
DEFAULT_RELEVANCE_WEIGHT = float(os.environ.get('MEMORY_RELEVANCE_WEIGHT', '0.40'))
DEFAULT_RECENCY_WEIGHT = float(os.environ.get('MEMORY_RECENCY_WEIGHT', '0.35'))
DEFAULT_HEBBIAN_WEIGHT = float(os.environ.get('MEMORY_HEBBIAN_WEIGHT', '0.15'))
DEFAULT_STRENGTH_WEIGHT = float(os.environ.get('MEMORY_STRENGTH_WEIGHT', '0.10'))

# In semantic mode, fetch more candidates for full ranking
_SEMANTIC_MODE = os.environ.get('VITRIOL_SEMANTIC_MODE', 'off').lower() == 'on'
_CANDIDATE_MULTIPLIER = 20 if _SEMANTIC_MODE else 10


def classify_intent(query: str) -> str:
    """Simple keyword-based intent classification."""
    debug_words = {'fix', 'bug', 'error', 'crash', 'broken', 'issue', 'fault', 'null'}
    question_words = {'how', 'what', 'why', 'when', 'where', 'explain', 'meaning', 'purpose'}
    create_words = {'add', 'implement', 'create', 'refactor', 'write', 'build', 'new'}

    q_lower = query.lower()
    q_words = set(re.findall(r'\w+', q_lower))

    if q_words & debug_words:
        return 'code_debug'
    if q_words & question_words:
        return 'question'
    if q_words & create_words:
        return 'code_write'
    return 'general'


def _merge_candidates(existing: list[dict], new: list[dict]) -> list[dict]:
    """Merge new candidates into existing list, de-duplicating by (type, id)."""
    seen = {(c.get('_type', 'episode'), c.get('id')) for c in existing}
    for candidate in new:
        key = (candidate.get('_type', 'episode'), candidate.get('id'))
        if key not in seen:
            existing.append(candidate)
            seen.add(key)
    return existing


def retrieve(
    project_id: str,
    query: str,
    top_k: int = DEFAULT_TOP_K,
    cascade_depth: int = DEFAULT_CASCADE_DEPTH,
    include_history: bool = False
) -> list[dict]:
    """
    Main retrieval pipeline.

    1. Hop 1: Direct search over episodes and knowledge nodes
    2. Hop 2+: Edge traversal (spreading activation)
    3. Score and rank
    include_history (2026-08-06 P3.2): include superseded node versions (default:
    current versions only).
    """
    candidates = []

    # ── Hop 1: Direct Retrieval ──
    # Use larger candidate pool in semantic mode for full ranking
    episodes = db.search_episodes(project_id, query, limit=top_k * _CANDIDATE_MULTIPLIER)
    for ep in episodes:
        ep['_type'] = 'episode'
        ep['_content'] = ep.get('content', '')
        ep['_source'] = 'hop1_direct'
    _merge_candidates(candidates, episodes)

    # In semantic mode, fetch all nodes (no pre-filtering needed)
    node_limit = top_k * (_CANDIDATE_MULTIPLIER // 3) if _SEMANTIC_MODE else top_k * 2
    nodes = db.search_nodes(project_id, query, limit=node_limit,
                            include_history=include_history)
    for n in nodes:
        n['_type'] = 'node'
        n['_content'] = n.get('summary', '')
        n['_source'] = 'hop1_direct'
    _merge_candidates(candidates, nodes)

    # ── Hop 2+: Cascading ──
    for depth in range(cascade_depth):
        hop_candidates = []
        for candidate in candidates:
            edges = db.get_outgoing_edges(
                project_id, candidate.get('_type', 'episode'), candidate['id']
            )
            for edge in edges:
                targets = db.get_edge_targets(
                    project_id, edge['from_type'], edge['from_id']
                )
                for target in targets:
                    if '_type' not in target:
                        target['_type'] = edge['to_type']
                    if '_content' not in target:
                        if 'content' in target:
                            target['_content'] = target['content']
                        elif 'summary' in target:
                            target['_content'] = target['summary']
                        else:
                            target['_content'] = ''
                    target['_source'] = f'hop{depth + 2}_cascade'
                    target['_edge_weight'] = edge.get('weight', 1.0)
                    target['_edge_relation'] = edge.get('relation', '')
                _merge_candidates(hop_candidates, targets)
        candidates = _merge_candidates(candidates, hop_candidates)

    # ── Score and Rank ──
    scored = []
    for candidate in candidates:
        content = candidate.get('_content', '')
        created_at = candidate.get('created_at')
        hebbian_w = candidate.get('_edge_weight', 0.5)
        strength = candidate.get('strength', 1.0)

        score = compute_score(
            query=query,
            content=content,
            created_at=created_at,
            hebbian_weight=hebbian_w,
            node_strength=strength,
            relevance_weight=DEFAULT_RELEVANCE_WEIGHT,
            recency_weight=DEFAULT_RECENCY_WEIGHT,
            hebbian_coeff=DEFAULT_HEBBIAN_WEIGHT,
            strength_coeff=DEFAULT_STRENGTH_WEIGHT,
        )
        candidate['_score'] = score
        # 2026-08-06 (P3.2): nodes are the current-world source; prefer them over
        # historical episodes when scores tie or are close.
        if candidate.get('_type') == 'node':
            candidate['_score'] = score + 0.05
        scored.append(candidate)

    scored.sort(key=lambda c: c['_score'], reverse=True)
    return scored[:top_k]


def context_block(project_id: str, recent_text: str,
                  budget_tokens: int = 1500, top_k: int = 5,
                  min_score: float = 0.3, session_id: str = None) -> tuple:
    """Build a budget-capped context block for selective injection.

    C (2026-08-06): feeds the Copula rolling-window auto-injection. Uses current node
    versions only (retrieve() defaults to superseded=0). Selective: drops candidates
    below min_score, returns (block, top_score, is_new_topic). is_new_topic is False when
    the query is semantically close to the recent session (a continuation the window
    already carries) — the plugin skips injecting then.
    """
    candidates = retrieve(project_id, recent_text, top_k=top_k, cascade_depth=1)
    candidates = [c for c in candidates if c.get('_score', 0.0) >= min_score]
    top_score = candidates[0]['_score'] if candidates else 0.0
    is_new = _is_new_topic(project_id, session_id, recent_text)
    lines = []
    used = 0
    for c in candidates:
        if c.get('_type') == 'node':
            body = compact.format_node(c)
        else:
            body = compact.format_episode(c)
        toks = estimate_tokens(body) + 1
        if used + toks > budget_tokens and used > 0:
            break
        used += toks
        lines.append(body)
    return '\n\n'.join(lines), top_score, is_new


def _is_new_topic(project_id: str, session_id: str, recent_text: str,
                  recent_n: int = 5, threshold: float = 0.55) -> bool:
    """True if the query is NOT a continuation of the session's recent turns.

    Embeds the query and the last recent_n episodes; low cosine -> new topic.
    Falls back to True (inject-worthy) when embeddings are unavailable.
    """
    if not session_id:
        return True
    from . import embed
    if not embed.is_available():
        return True
    q = embed.encode(recent_text[:1000])
    if q is None:
        return True
    conn = db._get_conn(project_id)
    rows = conn.execute(
        "SELECT content FROM episodes WHERE session_id=? ORDER BY id DESC LIMIT ?",
        (session_id, recent_n)).fetchall()
    if not rows:
        return True
    embs = [embed.encode(r['content'][:1000]) for r in rows]
    embs = [e for e in embs if e]
    if not embs:
        return True
    n, d = len(embs), len(embs[0])
    mean = [sum(embs[i][j] for i in range(n)) / n for j in range(d)]
    qn = math.sqrt(sum(x * x for x in q))
    mn = math.sqrt(sum(x * x for x in mean))
    if qn == 0.0 or mn == 0.0:
        return True
    cos = sum(q[i] * mean[i] for i in range(d)) / (qn * mn)
    return cos < threshold

File: libhermes/src/retrieval.rs
//! Retrieval with cascading (spreading activation) — port of
//! `hermetis/retrieval.py`. Direct search → edge traversal → score → rank.
//! Semantic-off path is bit-exact (keyword Jaccard); the semantic path calls
//! the same GPU embed server and is verified with tolerance.

use regex::Regex;
use std::collections::HashSet;

use crate::compact;
use crate::db::Hermes;
use crate::scorer::{compute_score, estimate_tokens, keyword_overlap};

/// Retrieval parameters (bundled; Python defaults + env knobs).
#[derive(Debug, Clone, Copy)]
pub struct RetrieveParams {
    pub top_k: usize,
    pub cascade_depth: usize,
    pub include_history: bool,
    /// Candidate multiplier: 20 semantic-on, 10 semantic-off (Python).
    pub candidate_multiplier: usize,
}

impl Default for RetrieveParams {
    fn default() -> Self {
        Self {
            top_k: env_usize("MEMORY_TOP_K", 5),
            cascade_depth: env_usize("MEMORY_CASCADE_DEPTH", 1),
            include_history: false,
            candidate_multiplier: 10,
        }
    }
}

fn env_usize(key: &str, default: usize) -> usize {
    std::env::var(key)
        .ok()
        .and_then(|v| v.parse().ok())
        .unwrap_or(default)
}

/// Simple keyword-based intent classification — Python `classify_intent`.
pub fn classify_intent(query: &str) -> &'static str {
    let words: HashSet<String> = word_regex()
        .find_iter(&query.to_lowercase())
        .map(|m| m.as_str().to_string())
        .collect();
    if has_any(
        &words,
        &[
            "fix", "bug", "error", "crash", "broken", "issue", "fault", "null",
        ],
    ) {
        return "code_debug";
    }
    if has_any(
        &words,
        &[
            "how", "what", "why", "when", "where", "explain", "meaning", "purpose",
        ],
    ) {
        return "question";
    }
    if has_any(
        &words,
        &[
            "add",
            "implement",
            "create",
            "refactor",
            "write",
            "build",
            "new",
        ],
    ) {
        return "code_write";
    }
    "general"
}

/// True when any word in the set is present in the list.
fn has_any(words: &HashSet<String>, list: &[&str]) -> bool {
    words.iter().any(|w| list.contains(&w.as_str()))
}

fn word_regex() -> &'static Regex {
    static RE: std::sync::OnceLock<Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| Regex::new(r"\w+").unwrap())
}

/// Merge `new` into `existing`, de-duplicating by (type, id) — Python
/// `_merge_candidates`. Mutates `existing` and returns it.
fn merge_candidates(existing: &mut Vec<serde_json::Value>, new: &[serde_json::Value]) {
    let mut seen: HashSet<(String, i64)> = existing.iter().map(|c| (ctype(c), cid(c))).collect();
    for candidate in new {
        let key = (ctype(candidate), cid(candidate));
        if seen.insert(key) {
            existing.push(candidate.clone());
        }
    }
}

fn ctype(c: &serde_json::Value) -> String {
    c.get("_type")
        .and_then(|t| t.as_str())
        .unwrap_or("episode")
        .to_string()
}

fn cid(c: &serde_json::Value) -> i64 {
    c.get("id").and_then(|i| i.as_i64()).unwrap_or(-1)
}

/// The main retrieval pipeline — Python `retrieve`.
pub fn retrieve(
    h: &Hermes,
    project_id: &str,
    query: &str,
    p: &RetrieveParams,
) -> Vec<serde_json::Value> {
    let mut candidates = hop1_direct(h, project_id, query, p);
    candidates = cascade(h, project_id, &mut candidates, p);
    let w = score_weights();
    let mut scored: Vec<serde_json::Value> = candidates
        .into_iter()
        .map(|mut c| {
            c["_score"] = serde_json::json!(score_one(query, &c, &w));
            c
        })
        .collect();
    scored.sort_by(|a, b| {
        let sa = a.get("_score").and_then(|s| s.as_f64()).unwrap_or(0.0);
        let sb = b.get("_score").and_then(|s| s.as_f64()).unwrap_or(0.0);
        sb.partial_cmp(&sa).unwrap_or(std::cmp::Ordering::Equal)
    });
    scored.truncate(p.top_k);
    scored
}

/// Hop 1: direct retrieval over episodes and knowledge nodes.
fn hop1_direct(
    h: &Hermes,
    project_id: &str,
    _query: &str,
    p: &RetrieveParams,
) -> Vec<serde_json::Value> {
    let mut candidates: Vec<serde_json::Value> = Vec::new();
    if let Ok(eps) = h.fetch_episodes(project_id, (p.top_k * p.candidate_multiplier) as i64) {
        let mut tagged = eps;
        for ep in &mut tagged {
            ep["_type"] = serde_json::json!("episode");
            ep["_content"] = ep.get("content").cloned().unwrap_or(serde_json::json!(""));
            ep["_source"] = serde_json::json!("hop1_direct");
        }
        merge_candidates(&mut candidates, &tagged);
    }
    if let Ok(mut ns) = h.fetch_nodes(project_id, (p.top_k * 2) as i64, p.include_history) {
        for n in &mut ns {
            n["_type"] = serde_json::json!("node");
            n["_content"] = n.get("summary").cloned().unwrap_or(serde_json::json!(""));
            n["_source"] = serde_json::json!("hop1_direct");
        }
        merge_candidates(&mut candidates, &ns);
    }
    candidates
}

/// Hops 2+: edge traversal (spreading activation). Each level is one loop so
/// the cascade stays Praetor-clean despite the algorithm being multi-pass.
fn cascade(
    h: &Hermes,
    project_id: &str,
    candidates: &mut Vec<serde_json::Value>,
    p: &RetrieveParams,
) -> Vec<serde_json::Value> {
    for depth in 0..p.cascade_depth {
        let hop = expand_all(h, project_id, candidates, depth);
        let snapshot = hop;
        merge_candidates(candidates, &snapshot);
    }
    candidates.clone()
}

/// Expand every candidate's outgoing edges into tagged targets.
fn expand_all(
    h: &Hermes,
    project_id: &str,
    candidates: &[serde_json::Value],
    depth: usize,
) -> Vec<serde_json::Value> {
    let mut hop: Vec<serde_json::Value> = Vec::new();
    for candidate in candidates {
        let added = expand_one(h, project_id, candidate, depth);
        merge_candidates(&mut hop, &added);
    }
    hop
}

/// Expand one candidate's outgoing edges.
fn expand_one(
    h: &Hermes,
    project_id: &str,
    candidate: &serde_json::Value,
    depth: usize,
) -> Vec<serde_json::Value> {
    let from_type = ctype(candidate);
    let from_id = cid(candidate);
    let edges = h
        .outgoing_edges(project_id, &from_type, from_id)
        .unwrap_or_default();
    let mut out: Vec<serde_json::Value> = Vec::new();
    for edge in edges {
        let weight = edge
            .get("weight")
            .cloned()
            .unwrap_or(serde_json::json!(1.0));
        let relation = edge
            .get("relation")
            .cloned()
            .unwrap_or(serde_json::json!(""));
        let targets = h
            .edge_targets(project_id, &from_type, from_id)
            .unwrap_or_default();
        let tagged: Vec<serde_json::Value> = targets
            .into_iter()
            .map(|mut t| {
                t["_type"] = t
                    .get("_type")
                    .cloned()
                    .unwrap_or(serde_json::json!("episode"));
                if t.get("_content").and_then(|x| x.as_str()).is_none() {
                    let content = t
                        .get("content")
                        .or_else(|| t.get("summary"))
                        .and_then(|x| x.as_str())
                        .unwrap_or("");
                    t["_content"] = serde_json::json!(content);
                }
                t["_source"] = serde_json::json!(format!("hop{}_cascade", depth + 2));
                t["_edge_weight"] = weight.clone();
                t["_edge_relation"] = relation.clone();
                t
            })
            .collect();
        merge_candidates(&mut out, &tagged);
    }
    out
}

/// The scoring weights (env-overridable).
fn score_weights() -> crate::scorer::ScoreWeights {
    crate::scorer::ScoreWeights {
        relevance_weight: env_f64("MEMORY_RELEVANCE_WEIGHT", 0.40),
        recency_weight: env_f64("MEMORY_RECENCY_WEIGHT", 0.35),
        hebbian_coeff: env_f64("MEMORY_HEBBIAN_WEIGHT", 0.15),
        strength_coeff: env_f64("MEMORY_STRENGTH_WEIGHT", 0.10),
    }
}

/// Score one candidate; nodes get a +0.05 current-world bonus (P3.2).
fn score_one(query: &str, candidate: &serde_json::Value, w: &crate::scorer::ScoreWeights) -> f64 {
    let content = candidate
        .get("_content")
        .and_then(|c| c.as_str())
        .unwrap_or("");
    let input = crate::scorer::ScoreInput {
        query,
        content,
        created_at: candidate.get("created_at").and_then(|c| c.as_str()),
        hebbian_weight: candidate
            .get("_edge_weight")
            .and_then(|w| w.as_f64())
            .unwrap_or(0.5),
        node_strength: candidate
            .get("strength")
            .and_then(|s| s.as_f64())
            .unwrap_or(1.0),
    };
    let mut score = compute_score(&input, w);
    if ctype(candidate) == "node" {
        score += 0.05;
    }
    score
}
fn env_f64(key: &str, default: f64) -> f64 {
    std::env::var(key)
        .ok()
        .and_then(|v| v.parse().ok())
        .unwrap_or(default)
}

/// Options for `context_block` (bundled to stay under the param gate).
#[derive(Debug, Clone)]
pub struct ContextBlockOptions {
    pub budget_tokens: usize,
    pub top_k: usize,
    pub min_score: f64,
    pub session_id: Option<String>,
}

/// Build a budget-capped context block — Python `context_block`. Semantic-off:
/// `is_new_topic` is always True (embeddings unavailable), so injection is
/// never skipped by the topic heuristic.
pub fn context_block(
    h: &Hermes,
    project_id: &str,
    recent_text: &str,
    o: &ContextBlockOptions,
) -> (String, f64, bool) {
    let params = RetrieveParams {
        top_k: o.top_k,
        cascade_depth: 1,
        include_history: false,
        candidate_multiplier: 10,
    };
    let mut candidates: Vec<serde_json::Value> = retrieve(h, project_id, recent_text, &params)
        .into_iter()
        .filter(|c| c.get("_score").and_then(|s| s.as_f64()).unwrap_or(0.0) >= o.min_score)
        .collect();
    let top_score = candidates
        .first()
        .and_then(|c| c.get("_score").and_then(|s| s.as_f64()))
        .unwrap_or(0.0);
    // Semantic-off: the embed server is unavailable, so Python `_is_new_topic`
    // returns True (inject-worthy) unconditionally.
    let is_new = true;
    let mut lines = Vec::new();
    let mut used = 0usize;
    for c in &mut candidates {
        let body = if ctype(c) == "node" {
            compact::format_node(c)
        } else {
            compact::format_episode(c, None)
        };
        let toks = estimate_tokens(&body) + 1;
        if used + toks > o.budget_tokens && used > 0 {
            break;
        }
        used += toks;
        lines.push(body);
    }
    (lines.join("\n\n"), top_score, is_new)
}

/// Keyword overlap exposed for cross-checks (semantic-off relevance).
pub fn keyword_overlap_pub(query: &str, content: &str) -> f64 {
    keyword_overlap(query, content)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{EpisodeMeta, EpisodeWrite};

    fn tmp_store(name: &str) -> (std::path::PathBuf, Hermes) {
        let root = std::env::temp_dir().join(format!("hermes_retrieval_{name}"));
        let _ = std::fs::remove_dir_all(&root);
        (root.clone(), Hermes::new(&root))
    }

    #[test]
    fn classify_intent_kinds() {
        assert_eq!(classify_intent("fix the crash in parser"), "code_debug");
        assert_eq!(classify_intent("how does the scheduler work"), "question");
        assert_eq!(classify_intent("implement a ring buffer"), "code_write");
        assert_eq!(classify_intent("ok"), "general");
    }

    #[test]
    fn retrieve_scores_and_ranks() {
        let (root, h) = tmp_store("ret");
        h.get_or_create_session("p", "s").unwrap();
        h.store_episode(
            "p",
            "s",
            &EpisodeWrite {
                role: "user".into(),
                content: "the red fox jumps over the lazy dog today quickly".into(),
                meta: EpisodeMeta::default(),
            },
        )
        .unwrap();
        h.store_episode(
            "p",
            "s",
            &EpisodeWrite {
                role: "assistant".into(),
                content: "the red fox is a mammal with a bushy tail".into(),
                meta: EpisodeMeta::default(),
            },
        )
        .unwrap();
        h.store_node(
            "p",
            "fox-note",
            "the red fox hunts at dusk and sleeps by day",
            &crate::NodeMeta {
                git_rev: "abc".into(),
                ..Default::default()
            },
        )
        .unwrap();
        let out = retrieve(&h, "p", "red fox", &RetrieveParams::default());
        assert!(!out.is_empty());
        // best match should be the node or the most-overlapping episode
        let best = &out[0];
        assert!(best["_score"].as_f64().unwrap() > 0.0);
        // node bonus: the node should outrank episodes with same overlap
        let node = out.iter().find(|c| ctype(c) == "node");
        assert!(node.is_some());
        let _ = std::fs::remove_dir_all(&root);
    }

    #[test]
    fn retrieve_cascades_edges() {
        let (root, h) = tmp_store("cascade");
        h.get_or_create_session("p", "s").unwrap();
        let a = h
            .store_episode(
                "p",
                "s",
                &EpisodeWrite {
                    role: "user".into(),
                    content: "alpha one two three four five".into(),
                    meta: EpisodeMeta::default(),
                },
            )
            .unwrap();
        let b = h
            .store_episode(
                "p",
                "s",
                &EpisodeWrite {
                    role: "assistant".into(),
                    content: "beta six seven eight nine ten".into(),
                    meta: EpisodeMeta::default(),
                },
            )
            .unwrap();
        // explicit cross edge a->b already exists (follows); query matches b only
        let out = retrieve(&h, "p", "beta six seven", &RetrieveParams::default());
        assert!(out.iter().any(|c| cid(c) == b));
        let _ = a;
        let _ = std::fs::remove_dir_all(&root);
    }

    #[test]
    fn context_block_budget_caps() {
        let (root, h) = tmp_store("ctx");
        h.get_or_create_session("p", "s").unwrap();
        h.store_episode(
            "p",
            "s",
            &EpisodeWrite {
                role: "user".into(),
                content: "the quick brown fox jumps".into(),
                meta: EpisodeMeta::default(),
            },
        )
        .unwrap();
        let (block, top, is_new) = context_block(
            &h,
            "p",
            "quick brown fox",
            &ContextBlockOptions {
                budget_tokens: 200,
                top_k: 5,
                min_score: 0.0,
                session_id: Some("s".into()),
            },
        );
        assert!(block.contains("fox") || block.is_empty());
        assert!(top >= 0.0);
        assert!(is_new); // semantic-off always new
        let _ = std::fs::remove_dir_all(&root);
    }
}
